// ng2-mobx/dist/ng2-mobx.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var core_1 = require("@angular/core");
var mobx_autorun_directive_1 = require("./directives/mobx-autorun.directive");
exports.MobxAutorunDirective = mobx_autorun_directive_1.MobxAutorunDirective;
var mobx_autorun_sync_directive_1 = require("./directives/mobx-autorun-sync.directive");
exports.MobxAutorunSyncDirective = mobx_autorun_sync_directive_1.MobxAutorunSyncDirective;
var mobx_reaction_directive_1 = require("./directives/mobx-reaction.directive");
exports.MobxReactionDirective = mobx_reaction_directive_1.MobxReactionDirective;
var DIRECTIVES = [mobx_autorun_directive_1.MobxAutorunDirective, mobx_autorun_sync_directive_1.MobxAutorunSyncDirective, mobx_reaction_directive_1.MobxReactionDirective];
var Ng2MobxModule = (function () {
    function Ng2MobxModule() {
    }
    return Ng2MobxModule;
}());
Ng2MobxModule = __decorate([
    core_1.NgModule({
        declarations: [
            DIRECTIVES
        ],
        exports: [
            DIRECTIVES
        ],
        imports: [],
        providers: []
    })
], Ng2MobxModule);
exports.Ng2MobxModule = Ng2MobxModule;
