// ng2-mobx/dist/utils/ng2-mobx-debug.js
"use strict";
var mobx_1 = require("mobx");
// function for turning debug on / off
window['ng2MobxDebug'] = function (value) {
    if (value)
        window['localStorage']['ng2-mobx-debug'] = true;
    else
        delete window['localStorage']['ng2-mobx-debug'];
};
function isDebugOn() {
    return window['localStorage'] && window['localStorage']['ng2-mobx-debug'];
}
mobx_1.spy(function (change) { return isDebugOn() && consoleLogChange(change, function () { return true; }); });
// Debugging element dependency tree
function ng2MobxDebug(view, renderer, observer) {
    if (!isDebugOn())
        return;
    var element = view.rootNodes[0];
    var debugElement = document.createElement('span');
    element.prepend(debugElement); // No prepend in Angular Renderer
    renderer.setElementStyle(debugElement, 'position', 'absolute');
    renderer.setElementStyle(debugElement, 'border', '1px dotted red');
    renderer.setElementStyle(debugElement, 'transform', "translateY(-25px)");
    renderer.setElementStyle(debugElement, 'cursor', 'pointer');
    renderer.setElementStyle(debugElement, 'z-index', '1000000');
    renderer.setElementStyle(debugElement, 'padding', '5px 10px');
    renderer.setElementStyle(debugElement, 'font-size', '14px');
    renderer.setElementStyle(debugElement, 'line-height', '14px');
    renderer.setElementStyle(debugElement, 'display', 'none');
    renderer.createText(debugElement, 'mobx deps');
    renderer.listen(element, 'mouseenter', function () {
        renderer.setElementStyle(debugElement, 'display', 'inline');
    });
    renderer.listen(element, 'mouseleave', function () {
        renderer.setElementStyle(debugElement, 'display', 'none');
    });
    renderer.listen(debugElement, 'click', function () {
        console.log(mobx_1.extras.getDependencyTree(observer));
    });
}
exports.ng2MobxDebug = ng2MobxDebug;
/////////////////////////////////////////////////////////
// console logging (copied from mobx-react)
var advicedToUseChrome = false;
var currentDepth = 0;
var isInsideSkippedGroup = false;
function consoleLogChange(change, filter) {
    if (advicedToUseChrome === false && typeof navigator !== 'undefined' && navigator.userAgent.indexOf('Chrome') === -1) {
        console.warn("The output of the MobX logger is optimized for Chrome");
        advicedToUseChrome = true;
    }
    var isGroupStart = change.spyReportStart === true;
    var isGroupEnd = change.spyReportEnd === true;
    var show;
    if (currentDepth === 0) {
        show = filter(change);
        if (isGroupStart && !show) {
            isInsideSkippedGroup = true;
        }
    }
    else if (isGroupEnd && isInsideSkippedGroup && currentDepth === 1) {
        show = false;
        isInsideSkippedGroup = false;
    }
    else {
        show = isInsideSkippedGroup !== true;
    }
    if (show && isGroupEnd) {
        groupEnd(change.time);
    }
    else if (show) {
        var logNext = isGroupStart ? group : log;
        switch (change.type) {
            case 'action':
                // name, target, arguments, fn
                logNext("%caction '%s' %s", 'color:dodgerblue', change.name, autoWrap("(", getNameForThis(change.target)));
                log(change.arguments);
                trace();
                break;
            case 'transaction':
                // name, target
                logNext("%ctransaction '%s' %s", 'color:gray', change.name, autoWrap("(", getNameForThis(change.target)));
                break;
            case 'scheduled-reaction':
                // object
                logNext("%cscheduled async reaction '%s'", 'color:#10a210', observableName(change.object));
                break;
            case 'reaction':
                // object, fn
                logNext("%creaction '%s'", 'color:#10a210', observableName(change.object));
                // dir({
                //     fn: change.fn
                // });
                trace();
                break;
            case 'compute':
                // object, target, fn
                group("%ccomputed '%s' %s", 'color:#10a210', observableName(change.object), autoWrap("(", getNameForThis(change.target)));
                // dir({
                //    fn: change.fn,
                //    target: change.target
                // });
                groupEnd();
                break;
            case 'error':
                // message
                logNext('%cerror: %s', 'color:tomato', change.message);
                trace();
                closeGroupsOnError();
                break;
            case 'update':
                // (array) object, index, newValue, oldValue
                // (map, obbject) object, name, newValue, oldValue
                // (value) object, newValue, oldValue
                if (mobx_1.isObservableArray(change.object)) {
                    logNext("updated '%s[%s]': %s (was: %s)", observableName(change.object), change.index, formatValue(change.newValue), formatValue(change.oldValue));
                }
                else if (mobx_1.isObservableObject(change.object)) {
                    logNext("updated '%s.%s': %s (was: %s)", observableName(change.object), change.name, formatValue(change.newValue), formatValue(change.oldValue));
                }
                else {
                    logNext("updated '%s': %s (was: %s)", observableName(change.object), change.name, formatValue(change.newValue), formatValue(change.oldValue));
                }
                dir({
                    newValue: change.newValue,
                    oldValue: change.oldValue
                });
                trace();
                break;
            case 'splice':
                // (array) object, index, added, removed, addedCount, removedCount
                logNext("spliced '%s': index %d, added %d, removed %d", observableName(change.object), change.index, change.addedCount, change.removedCount);
                dir({
                    added: change.added,
                    removed: change.removed
                });
                trace();
                break;
            case 'add':
                // (map, object) object, name, newValue
                logNext("set '%s.%s': %s", observableName(change.object), change.name, formatValue(change.newValue));
                dir({
                    newValue: change.newValue
                });
                trace();
                break;
            case 'delete':
                // (map) object, name, oldValue
                logNext("removed '%s.%s' (was %s)", observableName(change.object), change.name, formatValue(change.oldValue));
                dir({
                    oldValue: change.oldValue
                });
                trace();
                break;
            case 'create':
                // (value) object, newValue
                logNext("set '%s': %s", observableName(change.object), formatValue(change.newValue));
                dir({
                    newValue: change.newValue
                });
                trace();
                break;
            default:
                // generic fallback for future events
                logNext(change.type);
                dir(change);
                break;
        }
    }
    if (isGroupStart)
        currentDepth++;
    if (isGroupEnd)
        currentDepth--;
}
exports.consoleLogChange = consoleLogChange;
var consoleSupportsGroups = typeof console.groupCollapsed === "function";
var currentlyLoggedDepth = 0;
function group() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    // TODO: firefox does not support formatting in groupStart methods..
    consoleSupportsGroups ?
        console.groupCollapsed.apply(console, args) :
        console.log.apply(console, args);
    currentlyLoggedDepth++;
}
function groupEnd(time) {
    currentlyLoggedDepth--;
    if (typeof time === "number") {
        log("%ctotal time: %sms", "color:gray", time);
    }
    if (consoleSupportsGroups)
        console.groupEnd();
}
function log() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    console.log.apply(console, args);
}
function dir() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    console.dir.apply(console, args);
}
function trace() {
    // TODO: needs wrapping in firefox?
    console.trace("stack"); // TODO: use stacktrace.js or similar and strip off unrelevant stuff?
}
function closeGroupsOnError() {
    for (var i = 0, m = currentlyLoggedDepth; i < m; i++)
        groupEnd();
}
var closeToken = {
    "\"": "\"",
    "'": "'",
    "(": ")",
    "[": "]",
    "<": "]",
    "#": ""
};
function autoWrap(token, value) {
    if (!value)
        return "";
    return (token || "") + value + (closeToken[token] || "");
}
function observableName(object) {
    return mobx_1.extras.getDebugName(object);
}
function formatValue(value) {
    if (isPrimitive(value)) {
        if (typeof value === "string" && value.length > 100)
            return value.substr(0, 97) + "...";
        return value;
    }
    else
        return autoWrap("(", getNameForThis(value));
}
function getNameForThis(who) {
    if (who === null || who === undefined) {
        return "";
    }
    else if (who && typeof who === "object") {
        if (who && who.$mobx) {
            return who.$mobx.name;
        }
        else if (who.constructor) {
            return who.constructor.name || "object";
        }
    }
    return "" + typeof who;
}
function isPrimitive(value) {
    return value === null || value === undefined || typeof value === "string" || typeof value === "number" || typeof value === "boolean";
}
