// angular2-tree-component/dist/angular2-tree-component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TREE_ACTIONS } from './models/tree-options.model';
import { KEYS } from './constants/keys';
import { TreeModel } from './models/tree.model';
import { TreeNode } from './models/tree-node.model';
import { TreeDraggedElement } from './models/tree-dragged-element.model';
import { LoadingComponent } from './components/loading.component';
import { LoadingComponent as DeprecatedLoadingComponent } from './components/deprecated-loading.component';
import { TreeComponent } from './components/tree.component';
import { TreeNodeComponent } from './components/tree-node.component';
import { TreeNodeContent } from './components/tree-node-content.component';
import { TreeNodeContent as DeprecatedTreeNodeContent } from './components/deprecated-tree-node-content.component';
import { TreeNodeDropSlot } from './components/tree-node-drop-slot.component';
import { TreeNodeExpanderComponent } from './components/tree-node-expander.component';
import { TreeNodeChildrenComponent } from './components/tree-node-children.component';
import { TreeDropDirective } from './directives/tree-drop.directive';
import { TreeDragDirective } from './directives/tree-drag.directive';
import { AdHocComponentFactoryCreator } from './components/adhoc-component-factory.service';
import './polyfills';
import { deprecated } from './deprecated';
var exportedDirectives = [
    TreeComponent,
    TreeNodeComponent,
    TreeNodeContent,
    TreeDropDirective,
    TreeDragDirective,
    TreeNodeExpanderComponent,
    TreeNodeChildrenComponent,
    TreeNodeDropSlot
];
export { TreeModel, TreeNode, TreeDraggedElement, TREE_ACTIONS, KEYS, LoadingComponent, TreeComponent, TreeNodeComponent, TreeNodeContent, TreeDropDirective, TreeDragDirective, TreeNodeExpanderComponent, TreeNodeChildrenComponent, TreeNodeDropSlot };
var TreeModule = (function () {
    function TreeModule() {
    }
    return TreeModule;
}());
TreeModule = __decorate([
    NgModule({
        declarations: [
            LoadingComponent
        ].concat(exportedDirectives),
        exports: exportedDirectives.slice(),
        imports: [
            CommonModule,
        ],
        providers: [
            TreeDraggedElement
        ]
    })
], TreeModule);
export { TreeModule };
var DeprecatedTreeModule = (function () {
    function DeprecatedTreeModule() {
        deprecated('DeprecatedTreeModule', 'TreeModule for AoT compilation');
    }
    return DeprecatedTreeModule;
}());
DeprecatedTreeModule = __decorate([
    NgModule({
        declarations: [
            DeprecatedLoadingComponent,
            DeprecatedTreeNodeContent
        ],
        exports: [
            TreeComponent,
            TreeDropDirective,
            TreeDragDirective
        ],
        imports: [
            CommonModule,
            TreeModule,
        ],
        providers: [
            AdHocComponentFactoryCreator
        ],
    }),
    __metadata("design:paramtypes", [])
], DeprecatedTreeModule);
export { DeprecatedTreeModule };
export default TreeModule;
