// ng2-mobx/dist/directives/mobx-autorun.directive.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var mobx_1 = require("mobx");
var ng2_mobx_debug_1 = require("../utils/ng2-mobx-debug");
var MobxAutorunDirective = (function () {
    function MobxAutorunDirective(templateRef, viewContainer, renderer) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.renderer = renderer;
        this.templateBindings = {};
        this.view = this.viewContainer.createEmbeddedView(this.templateRef);
    }
    MobxAutorunDirective.prototype.ngAfterViewInit = function () {
        if (this.dispose)
            this.dispose();
        this.autoDetect(this.view);
        ng2_mobx_debug_1.ng2MobxDebug(this.view, this.renderer, this.dispose);
    };
    MobxAutorunDirective.prototype.autoDetect = function (view) {
        this.dispose = mobx_1.autorunAsync(function () {
            view["detectChanges"]();
        });
    };
    MobxAutorunDirective.prototype.ngOnDestroy = function () {
        if (this.dispose)
            this.dispose();
    };
    return MobxAutorunDirective;
}());
MobxAutorunDirective = __decorate([
    core_1.Directive({ selector: '[mobxAutorun]' }),
    __metadata("design:paramtypes", [core_1.TemplateRef,
        core_1.ViewContainerRef,
        core_1.Renderer])
], MobxAutorunDirective);
exports.MobxAutorunDirective = MobxAutorunDirective;
